src/cargo/sources/registry/mod.rs

```diff
@@ -198,29 +202,36 @@ pub struct RegistryConfig {
 }
 
 #[derive(Deserialize)]
-struct RegistryPackage {
-    name: String,
-    vers: String,
-    deps: Vec<RegistryDependency>,
+struct RegistryPackage<'a> {
+    name: Cow<'a, str>,
+    vers: Version,
+    deps: DependencyList,
     features: HashMap<String, Vec<String>>,
     cksum: String,
     yanked: Option<bool>,
 }
 
+struct DependencyList {
+    inner: Vec<Dependency>,
+}
+
 #[derive(Deserialize)]
-struct RegistryDependency {
-    name: String,
-    req: String,
+struct RegistryDependency<'a> {
+    name: Cow<'a, str>,
+    req: Cow<'a, str>,
     features: Vec<String>,
     optional: bool,
     default_features: bool,
-    target: Option<String>,
-    kind: Option<String>,
+    target: Option<Cow<'a, str>>,
+    kind: Option<Cow<'a, str>>,
 }
 
 pub trait RegistryData {
     fn index_path(&self) -> &Filesystem;
-    fn load(&self, root: &Path, path: &Path) -> CargoResult<Vec<u8>>;
+    fn load(&self,
+            _root: &Path,
+            path: &Path,
+            data: &mut FnMut(&[u8]) -> CargoResult<()>) -> CargoResult<()>;
```

Comment:
This could be make generic to return `CargoResult<T>`. Are we intentionally avoid monopolization here?